In [2]:
%matplotlib inline
import numpy as np
import random as rnd
import pandas as pd
import math
import matplotlib.pyplot as plt

__Задача 1__

Даны значения зарплат из выборки выпускников: 
```
100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 230, 24, 57, 55, 70, 75, 65, 84, 90, 150
```
Используя только встроенные питоновские функции и структуры данных (т.е. без библиотек `numpy`, `pandas` и др.), посчитайте (несмещённое) среднее квадратичное отклонение для данной выборки.

In [3]:
salaries = [100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 230, 24, 57, 55, 70, 75, 65, 84, 90, 150]

In [5]:
salaries_mean = sum(salaries) / len(salaries)

In [29]:
salaries_ = []
for i in salaries:
    salaries_.append((i - salaries_mean) ** 2)

var = sum(salaries_)/(len(salaries) - 1)

In [32]:
std = var ** 0.5

In [36]:
print(salaries_mean)
print(var)
print(std)

73.14285714285714
2241.8285714285716
47.34795213553139


Ответ: среднее квадратичное отклонение = 47.34795213553139

__Задача 2__

Найдите число выбросов в выборке из задачи 1. Для определения выбросов используйте методику как при построении "усов" в `boxplot`, однако, как и в задаче 1, пользоваться можно только встроенными функциями и структурами данных.

In [38]:
def quantile(l: list, order: float) -> float:
    
    l = sorted(l)
    n = len(l)
    
    n_left_elements = int(n * order)
    n_right_elements = int(n * (1 - order))
    
    print(n, n_left_elements, n_right_elements)
    
    if n_left_elements + n_right_elements == n:
        return (l[n_left_elements - 1] + l[n_left_elements]) / 2
    
    else:
        assert n_left_elements + n_right_elements == n - 1
        
        return l[n_left_elements]
    
    
def median(l: list) -> float:

    return quantile(l, 0.5)

In [42]:
q1 = quantile(salaries, 0.25)
q2 = np.median(salaries)
q3 = quantile(salaries, 0.75)
iqd = q3 - q1
boxplot_range = (q1 - 1.5 * iqd, q3 + 1.5 * iqd)
outliers = list(filter(lambda x: not (boxplot_range[0] <= x <= boxplot_range[1]), salaries))
print(outliers)

21 5 15
21 15 5
[230, 150]


__Задача 3__

В университет на факультет A поступило 100 человек, на факультет B - 90 человек и на факультет C - 45 человек. Вероятность того, что студент с факультета A сдаст первую сессию, равна 0.6. Для студента с факультета B эта вероятность равна 0.7, а для студента с факультета C - 0.8. Случайным образом выбирается студент, сдавший первую сессию. Какое событие наиболее вероятно:
1. студент учится на факультете A,
2. студент учится на факультете B,
3. студент учится на факультете C?

In [43]:
a = 100
b = 90
c = 45
total = a + b + c

p_a = a / total
p_b = b / total
p_c = c / total

p_a, p_b, p_c

(0.425531914893617, 0.3829787234042553, 0.19148936170212766)

In [45]:
p_sa = 0.5
p_sb = 0.6
p_sc = 0.7

p_full = p_sa * p_a + p_sb * p_b + p_sc * p_c
p_full

0.5765957446808511

In [60]:
p_as = p_sa * p_a / p_full
p_bs = p_sb * p_b / p_full
p_cs = p_sc * p_c / p_full

results = {'A':p_as , 'B':p_bs, 'C':p_cs}
max(results.items(), key = lambda k : k[1])

('B', 0.3985239852398524)